wrappers/python/simtk/openmm/app/modeller.py

```diff
@@ -1451,20 +1451,80 @@ def addMembrane(self, forcefield, lipidType='POPC', membraneCenterZ=0*nanometer,
-        # them with ions.
 
+        # Now add the ions. We do this by randomly selecting a set of waters a certain
+        # distance away from either leaflet of the membrane and replacing them with ions,
+        # while ensuring ions are not placed too close to each other.
         if totalIons > 0:
-            toReplace = random.sample(list(waterPos), totalIons)
+
+            # Redo waterPos because previous modeller.delete() changes Chain indexes
+            waterPos = {}
+            for chain in list(modeller.topology.chains())[-2:]:
+                for residue in chain.residues():
+                    if residue.name == 'HOH':
+                        for atom in residue.atoms():
+                            if atom.element == elem.oxygen:
+                                waterPos[residue] = modeller.positions[atom.index].value_in_unit(nanometer)
+
+            # Calculate lipid Z boundaries
+            lipidNames = {res.name for res in patch.topology.residues() if res.name != 'HOH'}
+            lipidZMax = 0.0
+            lipidZMin = 99999.0
+            for res in modeller.topology.residues():
+                if res.name in lipidNames:
+                    for atom in res.atoms():
+                        atomZ = modeller.positions[atom.index][2].value_in_unit(nanometer)
+                        lipidZMax = max(lipidZMax, atomZ)
+                        lipidZMin = min(lipidZMin, atomZ)
+
+            # Ignore waters that are within a certain distance of the membrane
+            lipidOffset = 0.25
+            upperZBoundary = (lipidZMax + lipidOffset)
+            lowerZBoundary = (lipidZMin - lipidOffset)
+            waterResidues = list(waterPos)
+            for wRes in waterResidues:
+                waterZ = waterPos[wRes][2]
+                if lowerZBoundary < waterZ < upperZBoundary:
+                    del waterPos[wRes]
+
+            # Select waters, ignoring those neighboring previously picked sites.
+            ionCutoff = 0.5
+            ionPositions = []
+
+            replaceableWaters = list(waterPos)
+
+            addedIons = 0
+            n_trials = 10
+            toReplace = []
+            while addedIons < totalIons:
+                pickedWater = random.choice(replaceableWaters)
+                replaceableWaters.remove(pickedWater)
+
+                # Check distance to other ions
+                for pos in ionPositions:
+                    distance = norm(pos - waterPos[pickedWater])
+                    if distance <= ionCutoff:
+                        n_trials -= 1
+                        break
+                else:
+                    toReplace.append(pickedWater)
+                    ionPositions.append(waterPos[pickedWater])
+                    addedIons += 1
+
+                    n_trials = 10
+
+                if n_trials == 0:
+                    raise Exception('Could not add more than {} ions to the system'.format(addedIons))
```

Comment:
Limiting it to 10 tries seems unnecessarily restrictive.  Every time through the loop you remove one water, so eventually it will either succeed or run out of waters.

Reply:
Also, it's probably better to use `ValueError` instead of `Exception`.  It's still kind of generic, but it basically means, "There was a problem with the value you specified."

Reply:
I removed the `n_trials` altogether and just catch and re-throw an exception when the `random.choice` hits an empty list.